### ISSUE TITLE
cmd/go,cmd/cgo: link-time symbol resolution for buildmode=c-shared

### ISSUE THREAD

**Issue by deweerdt** [golang/go#12216]
version:

```
$ go version
go version go1.5 linux/amd64
```

I'm working with a software that loads plugins structured as follows. Plugins implement an interface, and the main program provides an API, so a typical call stack might look like:

```
main_program_function()
\_ plugin_function()
   \_ main_program_other_function()
```

Up to now, the main program and the plugins are written in C. I was eager to see if I could start building plugins in Go instead of C, now that 1.5 is out, so I gave it a try with something that looks like:

``` go
package main

// #include "main_program_api.h"
import "C"

//export plugin_function
func plugin_function() {
   C.main_program_other_function()
}

func main () {
 // unused
}
```

Unfortunately, this fails with:

```
./plugin.go:8: undefined reference to `main_program_other_function'
```

Ian Lance Taylor suggested I pass '--unresolved-symbols' to the linker as a work around. I verified that that works as expected, both when passed via the command line:

```
CGO_LDFLAGS="-Wl,--unresolved-symbols=ignore-all" go build -buildmode=c-shared 
```

or via the code

``` go
// #cgo LDFLAGS: -Wl,--unresolved-symbols=ignore-all
```

**Comment by rsc** [comment]
It sounds like Ian fixed your problem. 

Are you proposing a change for the toolchain?

**Comment by deweerdt** [comment]
@rsc I do think that passing '--unresolved-symbols=ignore-all' would be a better defaut, but I can't say I feel strongly about it.

**Comment by rsc** [comment]
@ianlancetaylor, should we make -Wl,--unresolved-symbols=ignore-all the default for buildmode=c-shared uses of cgo?

If so, let's wait until Go 1.7 since the c-shared stuff is still experimental and there's an easy workaround. Otherwise, let's close the bug.

**Comment by ianlancetaylor** [comment]
The issue here isn't really with -buildmode=c-shared.  It's with using cgo to call a function that is not defined as far as cgo knows, because it is actually defined in a Go package.  This comes up pretty naturally with -buildmode=c-shared, but it's easy to construct a similar case in other build modes as well.

The real fix might be to have cgo do its links with -r, and not fuss about undefined symbols.  Or have cgo use --unresolved-symbols=ignore-all; I'm not sure which would work better.  Either way, we would let the final link give any errors about undefined symbols.

**Comment by ianlancetaylor** [comment]
The two options are completely different.  Using `-r` tells the linker to generate a relocatable object file.  Using `--unresolved-symbols=ignore-all` tells the linker to not report errors about undefined symbols.  The advantage of `-r` is that most linkers support it.  The advantage of `--unresolved-symbols=ignore-all` is that it does a normal link, not a relocatable link, so we don't have to deal with relocations.

For either one I'm not quite sure what will happen with an undefined symbol.  We may be left without any using DWARF info, which could be a problem in some cases.

**Comment by asottile** [comment]
I just stumbled on this thread while searching for the same thing (I'm attempting to build python modules written in go).  This solution was perfect, +1 for it becoming the default.

**Comment by sbinet** [comment]
@asottile it's only tangential but still a bit related: did you look at [gopy](https://github.com/go-python/gopy) to create python modules written in Go ?

**Comment by asottile** [comment]
Yeah it's not suitable for what I want to do.  You can see what I've done here though : https://github.com/asottile/setuptools-golang
